### PR TITLE
IGNITE-18843 Fix the behavior of MvPartitionStorage#pollForVacuum when trying to delete the same row in parallel

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-18843

### CODE REVIEW

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        Object lockedRowIds = this.lockedRowIds.get();
+
+        this.lockedRowIds.remove();
+
+        if (lockedRowIds == null) {
+            return;
+        } else if (lockedRowIds.getClass() == RowId.class) {
+            releaseLock0(((RowId) lockedRowIds), true);
+        } else {
+            ((Set<RowId>) lockedRowIds).forEach(rowId -> releaseLock0(rowId, true));
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java:138]: Fix it

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            this.lockedRowIds.remove();
+        } else {
+            Set<RowId> rowIds = ((Set<RowId>) lockedRowIds);
+
+            boolean remove = rowIds.remove(rowId);
+
+            assert remove : "rowId=" + rowId + ", lockedRowIds=" + rowIds;
+
+            if (rowIds.isEmpty()) {
+                this.lockedRowIds.remove();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java:118]: fix it

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+     * @param rowId Row ID.
+     */
+    public void acquireLock(RowId rowId) {
+        acquireLock0(rowId);
+
+        Object lockedRowIds = this.lockedRowIds.get();
+
+        if (lockedRowIds == null) {
+            this.lockedRowIds.set(rowId);
+        } else if (lockedRowIds.getClass() == RowId.class) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java:68]: fix it

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+            reentrantLockLockHolder.incrementHolders();
+
+            return reentrantLockLockHolder;
+        });
+
+        lockHolder.getLock().lock();
+    }
+
+    private void releaseLock0(RowId rowId, boolean untilHoldByCurrentThread) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java:156]: Tried to rename.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+        do {
+            lock.unlock();
+
+            lockHolderByRowId.compute(rowId, (rowId1, reentrantLockLockHolder) -> {
+                assert reentrantLockLockHolder != null;
+
+                return reentrantLockLockHolder.decrementHolders() ? null : reentrantLockLockHolder;
+            });
+        } while (untilHoldByCurrentThread && lock.getHoldCount() > 0);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java:173]: I think that I delete holder when `LockHolder#decrementHolders` returned `true` i.e. no one else holds the lock.

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        RowId rowId = new RowId(0);
+
+        lockByRowId.acquireLock(rowId);
+        lockByRowId.releaseLock(rowId);
+
+        assertEquals(1, lockByRowId.inLock(rowId, () -> 1));
+    }
+
+    @Test
+    void testSimpleReEnter() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java:49]: This is a simple test that checks that there are no errors when using the api.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvPartitionStorageConcurrencyTest.java [hunk 7]
```diff
@@ -193,19 +202,39 @@ void testConcurrentGc(AddAndCommit addAndCommit) {
...
             assertNull(storage.closestRowId(ROW_ID));
+
+            assertThat(rows, empty());
         }
     }
 
+    private void assertRemoveRow(byte[] rowBytes, Collection<byte[]> rows) {
+        assertNotNull(rowBytes);
+
+        byte[] found = rows.stream().filter(bytes -> Arrays.equals(bytes, rowBytes)).findFirst().orElseGet(null);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvPartitionStorageConcurrencyTest.java:227]: Fix it

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+        lockByRowId.releaseLock(rowId);
+
+        assertThat(acquireLockFuture, willFailFast(TimeoutException.class));
+
+        lockByRowId.releaseLock(rowId);
+
+        assertThat(acquireLockFuture, willCompleteSuccessfully());
+
+        lockByRowId.acquireLock(rowId);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java:106]: > Why do you re-acquire the lock?
Just to make sure we can acquire the lock again after another thread has used it.

> Also, why don't you use try/finally blocks in your tests?
I don’t think that it is necessary in these tests, they will only create unnecessary noise in the code.

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        CompletableFuture<?> acquireLockFuture = runAsync(() -> {
+            lockByRowId.acquireLock(rowId);
+            lockByRowId.releaseLock(rowId);
+        });
+
+        assertThat(acquireLockFuture, willFailFast(TimeoutException.class));
+
+        lockByRowId.releaseLock(rowId);
+
+        assertThat(acquireLockFuture, willFailFast(TimeoutException.class));
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java:100]: Tried to fix it

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java]: I'm sorry, I made a mistake, I probably meant a thread-local set. You nullify it once somewhere later, right?

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java:105]: I don't feel comfortable when there are unreleased locks in general. Can you please fix it? At least call that "release everything" method in "afterEach"

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/matchers/CompletableFutureMatcher.java [hunk 10]
```diff
@@ -174,6 +174,26 @@ public static CompletableFutureMatcher<Object> willFailIn(int time, TimeUnit tim
         return new CompletableFutureMatcher<>(anything(), time, timeUnit, cause);
     }
 
+    /**
+     * Creates a matcher that matches a future that <strong>not</strong> completes decently fast.
+     *
+     * @return matcher.
+     */
+    public static CompletableFutureMatcher<Object> willTimeoutFast() {
+        return willTimeoutIn(1, TimeUnit.SECONDS);
```
Comment (ibessonov) [inline modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/matchers/CompletableFutureMatcher.java:183]: 1 second is "fast"?

Comment (tkalkirill) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/matchers/CompletableFutureMatcher.java:183]: I can do a quarter/half of a second, for example, for the win os, less than 200 ms may not be enough.

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java:48]: I do it outside the method.

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/matchers/CompletableFutureMatcher.java (hunk 10 above)
Comment (tkalkirill) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/matchers/CompletableFutureMatcher.java:183]: set 250 ms

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowIdTest.java:105]: Try to fix it

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/ReentrantLockByRowId.java]: I do it outside the method.

**Review by ibessonov — APPROVED** [review]

**Pull request merged**